### DISCUSSION TITLE
[PROPOSAL] Add prompt argument types to the spec

### DISCUSSION THREAD

**Discussion by kdawgwilk** [modelcontextprotocol/modelcontextprotocol#68] [Ideas - General]
As I was reading the spec as it relates to prompts I was curious on what types of arguments could be utilized for prompts.  The example in the spec only shows a string argument type but it seems like would be useful to support other types of arguments.  For example a `string[]`.  I propose that we add some way to indicate on `PromptArgument` what the argument type is.  This could start with just a `string` type for now but would allow for supporting other types in the future e.g. `string[]`.

**Comment by dsp-ant** [comment]
I understand the concern. I'd love to expand the protocol based on very specific use-cases where the current implementation causes a lot of pain. If you have specific problems with the current implementations, let us know. We internally found the current representation good enough in nearly all cases, which of course doesn't mean there aren't any usecases. Would just love to make it specific rather than theoretical.

**Reply by kdawgwilk** [reply]
To be more concrete about this, we currently use [handlebars](https://handlebarsjs.com/) templates for our prompt templates and then define a corresponding [zod schema](https://zod.dev/) that represents the arguments that can be injected into the handlebars template string.  We do a fair amount of `{{#each someList}}` to render a list of items into a prompt template.  I currently do not see how that would be possible with the spec in its current state.  I guess you could pass a json stringified object or array?

```ts
const templateString = `
Some text for the prompt template
{{#each listOfObjects}}
  Render each item in the list {{this.name}}
{{/each}}
`
const templateSchema = z.object({
  listOfObjects: z.array(z.object({ name: z.string() })
})
```

**Comment by evalstate** [comment]
It looks like having the Client support completions https://spec.modelcontextprotocol.io/specification/server/utilities/completion/ would help here - perhaps if there were a way to indicate that completions should be preferred for population?

**Comment by localden** [comment]
We've limited Discussions in this repository to meeting notes and are closing this thread. To continue the conversation, pick a channel from [Contributor Communication](https://modelcontextprotocol.io/community/communication). Thanks for posting.
